File: java/fenwicktree.java
class Fenwicktree{
    private int n;
    private long[] bit;
    private long[] arr;

    public Fenwicktree(int n){
        this.n = n;
        bit = new long[n];
        arr = new long[n];
    }

    public void add(int i, long x){
        arr[i] += x;
        while (i < n){
            bit[i] += x;
            i |= i+1;
        }
    }

    public void update(int i, long val){
        long diff = val-arr[i];
        add(i, diff);
    }

    public long get(int i){
        return arr[i];
    }
    
    public long sum(int i){
        long res = 0;
        while (i >= 0){
            res += bit[i];
            i = (i&(i+1))-1;
        }
        return res;
    }

    public long rangeSum(int l, int r){
        if (l > r) return 0;
        return sum(r)-(l > 0? sum(l-1): 0);
    }
}

File: py/fenwicktree.py
class fenwicktree:
    def __init__(self, n):
        self.n = n
        self.bit = [0]*n
        self.a = [0]*n
    
    def add(self, i, x):
        while i < self.n:
            self.bit[i] += x
            i |= i+1
    
    def set(self, i, x):
        self.add(i, x-self.a[i])
        self.a[i] = x
    
    def get(self, i):
        return self.a[i]
    
    def sum(self, i):
        res = 0
        while i >= 0:
            res += self.bit[i]
            i = (i&(i+1))-1
        return res
    
    def range_sum(self, l, r):
        if l > r:
            return 0
        return self.sum(r)-(self.sum(l-1) if l else 0)